bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/grpc/RefreshingOAuth2CredentialsInterceptor.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright 2014 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.grpc;
+
+import io.grpc.Call;
+import io.grpc.Channel;
+import io.grpc.ClientInterceptor;
+import io.grpc.ClientInterceptors.CheckedForwardingCall;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+
+import java.io.IOException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import com.google.api.client.util.Clock;
+import com.google.api.client.util.Preconditions;
+import com.google.auth.oauth2.AccessToken;
+import com.google.auth.oauth2.OAuth2Credentials;
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Client interceptor that authenticates all calls by binding header data provided by a credential.
+ * Typically this will populate the Authorization header but other headers may also be filled out.
+ *
+ * <p> Uses the new and simplified Google auth library:
+ * https://github.com/google/google-auth-library-java</p>
+ *
+ * <p> TODO: COPIED FROM io.grpc.auth.ClientAuthInterceptor.  The logic added here for
+ * initialization and locking could be moved back to gRPC.  This implementation takes advantage
+ * of the fact that all of the Bigtable endpoints are OAuth2 based.  It uses the OAuth AccessToken
+ * to get the token value and next refresh time.  The refresh is scheduled asynchronously.</p>
+ */
+public class RefreshingOAuth2CredentialsInterceptor implements ClientInterceptor {
+
+  private enum CacheState {
+    Good,
+    Stale,
+    Expired
+  }
+
+  private static class TokenCache {
+    final IOException exception;
+    final String header;
+    final long staleTime;
+    final long expiresTime;
+
+    public TokenCache(String cachedHeader, long staleTime, long expiresTime) {
+      super();
```

Comment:
No reason.  It's always implied by default, but can be called explicitly.  Eclipse constructor generation adds them in for some reason.